tidy3d/exceptions.py

```diff
@@ -0,0 +1,60 @@
+"""Custom Tidy3D exceptions"""
+
+import logging
+
+
+# use root logger
+_log = logging.getLogger("tidy3d")
+
+
+class Tidy3dError(Exception):
+    """Any error in tidy3d"""
+
+    def __init__(self, message: str = None):
+        """Log just the error message and then raise the Exception."""
+        super().__init__(message)
+        _log.error(message)
+
+
+class ConfigError(Tidy3dError):
+    """Error when configuring Tidy3d."""
+
+
+class Tidy3dKeyError(Tidy3dError):
+    """Could not find a key in a Tidy3d dictionary."""
+
+
+class ValidationError(Tidy3dError):
+    """Error when constructing Tidy3d components."""
+
+
+class SetupError(Tidy3dError):
+    """Error regarding the setup of the components (outside of domains, etc)."""
+
+
+class FileError(Tidy3dError):
+    """Error reading or writing to file."""
+
+
+class WebError(Tidy3dError):
+    """Error with the webAPI."""
+
+
+class AuthenticationError(Tidy3dError):
+    """Error authenticating a user through webapi webAPI."""
+
+
+class DataError(Tidy3dError):
+    """Error accessing data."""
+
+
+class Tidy3dImportError(Tidy3dError):
+    """Error importing a package needed for tidy3d."""
+
+
+class Tidy3dNotImplementedError(Tidy3dError):
+    """Error when a functionality is not (yet) supported."""
+
+
+class AdjointError(Tidy3dError):
```

Comment:
I originally had this in the `tidy3d/plugins/adjoint/log.py` file. The point was trying to separate the plugins (eg. adjoint) from the main package as much as possible so that the main package does not contain any plugin-specific code. 

Is it possible to move this into a `tidy3d/plugins/adjoint/exceptions.py` file to retain this separation? Or does it break anything you've done here?

Also, if you think it makes more sense to instead define all the exceptions (plugins included) in one file at the `tidy3d/exceptions.py` let me know as well.

Reply:
Yes, I should have mentioned this specifically. I though it would bake more sense to have all exceptions in one place, because it makes it easier for the user. Imagine they want to catch and treat exceptions in their own programs. Its easier to `import tidy3d.exceptions as tde` and have all Tidy3D-related exception under a single namespace than having to import from multiple modules. But I don't mind taking the definition back into the submodule if you want to.